File: cli/dispatch.rs
// Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
use crate::isolate_state::*;
use crate::ops;
use deno::deno_buf;
use deno::Dispatch;
use deno::Op;
use std::sync::Arc;

/// Implements deno::Dispatch for the main Deno command-line.
pub struct CliDispatch {
  pub state: Arc<IsolateState>,
}

impl CliDispatch {
  pub fn new(state: Arc<IsolateState>) -> Self {
    Self { state }
  }
}

impl Dispatch for CliDispatch {
  fn dispatch(
    &mut self,
    control: &[u8],
    zero_copy: deno_buf,
  ) -> (bool, Box<Op>) {
    ops::dispatch_all(&self.state, control, zero_copy, ops::op_selector_std)
  }
}

File: cli/js/dispatch.ts
// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
import * as minimal from "./ops/dispatch_minimal.ts";
import * as json from "./ops/dispatch_json.ts";
import { AsyncHandler } from "./plugins.ts";

const PLUGIN_ASYNC_HANDLER_MAP: Map<number, AsyncHandler> = new Map();

export function setPluginAsyncHandler(
  opId: number,
  handler: AsyncHandler
): void {
  PLUGIN_ASYNC_HANDLER_MAP.set(opId, handler);
}

export function getAsyncHandler(opName: string): (msg: Uint8Array) => void {
  switch (opName) {
    case "op_write":
    case "op_read":
      return minimal.asyncMsgFromRust;
    default:
      return json.asyncMsgFromRust;
  }
}
